gradle/project.gradle.kts

```diff
@@ -177,29 +180,40 @@ tasks {
             val kotlincUpToDate = kotlinc.resolve("build.txt").takeIf { it.exists() }?.readText()?.startsWith(kotlinVersion)==true
             if (!kotlincUpToDate) {
                 if (kotlinc.exists()) {
-                    println("Previous version of Kotlin compiler exists. Deleting...")
+                    println("Deleting obsolete version of Kotlin compiler...")
                     if (!kotlinc.deleteRecursively())
                         throw IOException("Failed to remove Kotlin compiler, location=$kotlinc")
                 }
 
                 try {
                     val url = URL("https://github.com/JetBrains/kotlin/releases/download/v$kotlinVersion/kotlin-compiler-$kotlinVersion.zip")
-                    val zip = ZipInputStream(url.openStream())
+                    val con = url.openConnection()
+                    val zip = ZipInputStream(con.getInputStream())
+                    val max = con.contentLengthLong / 1000
+                    var downloaded = 0L
+                    var last = ""
                     while (true) {
                         val next = zip.nextEntry ?: break
+                        downloaded += next.size / 1000
                         if (!next.isDirectory) {
-                            val file = workDir.resolve(next.name)
+                            val file = if (next.name=="kotlinc/build.txt") buildDir.resolve("build.txt")
+                            else workDir.resolve(next.name).also {
+                                if (last!=it.parent) {
+                                    last = it.parent
+                                    println("Downloading ${it.parentFile.relativeTo(workDir)} downloaded: $downloaded/${max}k")
+                                }
+                            }
                             file.parentFile.mkdirs()
                             val out = file.outputStream()
                             zip.copyTo(out)
-                        } else {
-                            println("Downloading ${next.name}")
                         }
                     }
+                    zip.close()
                     if (!kotlinc.exists())
                         throw IOException("Kotlinc has not been downloaded successfully! Maybe the remote file is not a zip?")
 
                     file("$workDir/kotlinc/bin/kotlinc").setExecutable(true) // Allow kotlinc to be executed on Unix
+                    buildDir.resolve("build.txt").renameTo(kotlinc.resolve("build.txt"))
```

Comment:
it is a safety net, because if the kotlinc download is interrupted, it was previously left in an undetermined state. Now the task will simply redownload it.

Reply:
How does renaming the file help the issue?

Reply:
the file that tells the task that kotlinc is already downloaded is now stored somewhere else and only moved into the correct place after the download finished